### PR TITLE
WIP Remove charmstore dep stage 4

### PR DESCRIPTION
**Pull request by timClicks**
Work in progress (tests not quite passing yet)
----

## Description of change

*Please replace with a description about why this change is needed?*

## QA steps

*Please replace with how we can verify that the change works?*

## Documentation changes

*Please replace with any notes about how it affects current user workflow? CLI? API?* 

## Bug reference

*Please add a link to any bugs that this change is related to.*

### CODE REVIEW

File: charmstore/fakeclient.go [hunk 1]
```diff
@@ -349,6 +349,9 @@ func NewRepository() *Repository {
 
 func (r *Repository) addRevision(ref *charm.URL) *charm.URL {
 	revision := r.revisions[r.channel][*ref]
+	if revision == 0 {
```
Comment (wallyworld) [inline charmstore/fakeclient.go]: Shouldn't need to do this...

File: charmstore/fakeclient.go [hunk 2]
```diff
@@ -421,7 +426,7 @@ func (r Repository) CharmInfo(charmURL string) (*charms.CharmInfo, error) {
 //
 // Part of the charmrepo.Interface
 func (r Repository) Resolve(ref *charm.URL) (canonRef *charm.URL, supportedSeries []string, err error) {
-	return r.addRevision(ref), []string{"trusty", "wily", "quantal"}, nil
+	return r.addRevision(ref), []string{"trusty", "wily", "quantal", "kubernetes"}, nil
```
Comment (wallyworld) [inline charmstore/fakeclient.go]: Can't have charms operating as both k8s and non k8s, or WIndows and Ubuntu etc
All series for a given charm need to be the same OS

File: cmd/juju/application/deploy.go [hunk 3]
```diff
@@ -88,13 +90,15 @@ type CharmDeployAPI interface {
 	CharmInfo(string) (*apicharms.CharmInfo, error)
 }
 
+type charmDeployClient struct {
```
Comment (wallyworld) [inline cmd/juju/application/deploy.go]: This isn't used anywhere?

File: cmd/juju/application/deploy.go [hunk 4]
```diff
@@ -137,11 +137,17 @@ type modelConfigClient struct {
 	*modelconfig.Client
 }
 
+type addCharmClient struct {
```
Comment (wallyworld) [inline cmd/juju/application/deploy.go]: This looks like only used in export_test, so move it there. Is it needed?

File: cmd/juju/application/deploy_test.go [hunk 5]
```diff
@@ -706,7 +708,9 @@ func (s *CAASDeploySuite) TestDevices(c *gc.C) {
 	c.Assert(err, jc.ErrorIsNil)
 
 	_, ch := testcharms.UploadCharmWithSeries(c, s.client, "kubernetes/bitcoin-miner-1", "bitcoin-miner", "kubernetes")
-	err = runDeploy(c, "bitcoin-miner", "-m", m.Name(), "--device", "bitcoinminer=10,nvidia.com/gpu", "--series", "kubernetes")
+
+	logger.Infof("charmrepo state %#v", s.charmrepo)
```
Comment (wallyworld) [inline cmd/juju/application/deploy_test.go]: Debugf

File: cmd/juju/application/deploy_test.go [hunk 6]
```diff
@@ -1021,6 +1018,34 @@ func (s *charmstoreSuite) assertDeployedApplicationBindings(c *gc.C, info map[st
 	}
 }
 
+// assertRelationsEstablished checks that the given relations have been set.
+func (s *charmstoreSuite) assertRelationsEstablished(c *gc.C, relations ...string) {
```
Comment (wallyworld) [inline cmd/juju/application/deploy_test.go]: Just check we need both of these new functions - we need one now, maybe we will need both eventually?

File: cmd/juju/application/deploy_test.go [hunk 7]
```diff
@@ -1034,140 +1059,16 @@ func (s *charmstoreSuite) runDeploy(c *gc.C, args ...string) error {
...
-
-	err = runDeploy(c, "--channel", "edge", "~client-username/wordpress")
-	c.Assert(err, gc.IsNil)
-	s.assertCharmsUploaded(c, "cs:~client-username/precise/wordpress-0")
-	s.assertApplicationsDeployed(c, map[string]applicationInfo{
-		"wordpress": {charm: "cs:~client-username/precise/wordpress-0", config: ch.Config().DefaultSettings()},
-	})
-}
+// TODO
+//func (s *DeployCharmStoreSuite) TestDeployWithChannel(c *gc.C) {
```
Comment (wallyworld) [inline cmd/juju/application/deploy_test.go]: Let's look to re-implement this as discussed live

**Comment by timClicks** [comment]
This all sounds sensible

File: cmd/juju/application/deploy.go (hunk 4 above)
Comment (timClicks) [reply cmd/juju/application/deploy.go]: Was following the pattern that's already introduced in that file with everything being wrapped in a local `struct`. Probably isn't required,

File: cmd/juju/application/deploy_test.go (hunk 6 above)
Comment (timClicks) [reply cmd/juju/application/deploy_test.go]: These were copied verbatim from `legacyCharmStoreSuite`. They're there to maintain current test behaviour.

File: cmd/juju/application/deploy_test.go (hunk 7 above)
Comment (timClicks) [reply cmd/juju/application/deploy_test.go]: 100%

**Comment by nammn** [comment]
How do we want to proceed with this?

**Pull request closed without merging**